_nodes/0006.md

```diff
@@ -1,6 +1,5 @@
 ---
 title: "Example: the family fibration"
-layout: page
```

Comment:
Yes! You can set front matter defaults by file path and collection type -- see [details here](https://jekyllrb.com/docs/configuration/front-matter-defaults/).

Reply:
Ah but I think this doesn't exactly do what I need --- In general I may have multiple root "lectures" that import different nodes; each lecture may use a different macro library. Do you see how to accommodate this usage, or do you have a suggestion for another way for me to organize my notes?

Reply:
(It is also OK to say "Don't have multiple macro libraries!!". I could make peace with that potentially.)

Reply:
Oh ok, I misunderstood your question.

Yes, we can simply use [this iteration](https://github.com/jonsterling/math/blob/d42ab9f8b56b243414c431312cdd798141f0280e/_plugins/sheafy.rb#L90) to push front matter values down from roots to leaves in a single pass.

The only issue I see with that is with children nodes belonging to multiple parents; that's not happening in this use case, but it's also not explicitly forbidden and might cause unexpected behaviour.

Reply:
Cool, thanks. My take is that both scenarios are badly behaved --- in particular, a node that depends on `macrolib1` cannot be rendered under any circumstances alongside a node that depends on `macrolib2`, because the KaTeX architecture imports macros for the entire page and not for specific regions of the page. (Maybe this could be addressed but I think it is asking for suffering, and not likely to satisfy many interesting use-cases). This is all to say that I think we should expect bad behavior in this case.

My original plan was to have one "math website" with a lot of different lectures/mini-books in it. Each of these would naturally require different macro-libraries. But it may be that this style of working is bad, and that one should instead imagine having one "sheafy website" per mini-book. In that case it would be natural to fix a global macro library for the entire thing.  This might be simpler for people to understand and probably would not rule out many real applications. 

What do you think?

Reply:
Ok, let me spell things out:

- a node must be renderable as a standalone;
- a node must be renderable when embedded;
- a node must always yield the same result when rendered.

So basically embedding/rendering must compositionally well-behaved.
This greatly restricts what we factors altering rendering (or embedding) can do.

In particular, it implies that a node must always have the same macro set of any of its parents, and therefore of any of its ancestors.

It seems like that setting this property only on ancestors and letting it dribble down (like I suggested in the previous message) should be the correct thing to do.

However, this breaks when a node is a descendant of two nodes which have no common ancestor: if the latter have different properties, an ambiguity happens.

Forbidding this situation is equivalent to asking that each connected component of the DAG is rooted, i.e. it hase a base node which is an ancestor of every other one therefore fixing the property on the component.

We could even forbid reusing nodes, therefore making the DAG a **Directed Rooted Forest**.
Should we? I can think of three reasons:

- hyperlinking from the TOC to the anchors of different in-page instances of the same node would be annoying
- reasoning about trees would be easier for the user
- rooted trees are exactly what `gerby` produces from LaTeX documents

TL;DR: we should check the DAG is a **DRF** and let some properties dribble down the trees; this allows you to maintain multiple separate mini-books within the same Jekyll installation in a very natural way.

Reply:
Thanks for this, sorry for the delay in getting back to you! I agree with your analysis of the situation, and come to the same conclusions as you. In particular, I think that it actually makes a lot of sense to enforce DRFs; this simplifies a lot of things while only ruling out what I would consider to be pathological use-cases of the tool. In general, while we should be in principle open to deviating from the Gerby model, Gerby also provides a wealth of experience about what is and is not useful. So I like and endorse your idea. Thanks!

Reply:
Great! As you saw I prepared a nest at https://github.com/paolobrasolin/jekyll-sheafy
All topological checks are implemented in https://github.com/paolobrasolin/jekyll-sheafy/blob/main/lib/jekyll/sheafy/directed_graph.rb
There's a PR to use them in `math`: https://github.com/jonsterling/math/pull/13